### PR TITLE
fix: Repeating events in ld+json that share a common URL no longer overwrite each other

### LINKED ISSUE #2553: [Bug]: Outsavvy repeating events only showing the one farthest in future
## Description

What were you doing?

## Steps to reproduce

1. Add OutSavvy calendar with event series data
2. Only see the last event in each series in the Admin display
3. Need to verify whether we also only see the last event in the public listsing

## What you expected to happen

Display all the events in the series in both views.
In the admin view, an explanation might be good enough rather than display recurring.

## Platform (if relevant)

Raised in discord by Fruit Salad Events organiser "Hello, one of our partners uses recurring events on Outsavvy - it looks like only the latest event in the series is showing for each listing:"

image

### PR DESCRIPTION
**Pull request by kimadactyl**
Started a bit of work on this, it's taking a fair bit of refactoring for me to make sense of what's going on.

This PR:

- Fixes ld+json repeating events with a shared URL not overwriting each other
- Renames the `import` rake namespace to `event` to match the file name and given them more legible names. The mismatch here was very confusing for me, but this will require updating the cron job.
- Adds proper descriptions to these tasks so they show up properly when you run `rails --tasks`
- Silences some messages about implicit gem dependancies
- Refactors the ld+json importer a bit
- Removes a superflous rake task
- Refactors the ld+json importer to make it more legible, including marking protected methods properly
- Adds a chunk about how to use the importer into the README as right now this isn't documented anywhere

Fixes #2553

### CODE REVIEW

File: app/jobs/calendar_importer/parsers/ld_json.rb [hunk 1]
```diff
@@ -55,83 +55,19 @@ class EventConsumer
...
-
-      def consume_visual_arts_event(data)
-        consume_event data
+      event_types.each do |type|
+        define_method(:"consume_#{type}_event") do |data|
+          consume_event data
+        end
       end
 
-      def consume_course_instance(data); end
```
Comment (kimadactyl) [inline app/jobs/calendar_importer/parsers/ld_json.rb:126]: Very unclear why these are not implemented, assume this was an oversight but making a comment so someone can double check

File: app/jobs/calendar_importer/parsers/ld_json.rb [hunk 2]
```diff
@@ -55,83 +55,19 @@ class EventConsumer
...
-      def consume_event_series(data); end
-      def consume_festival(data); end
-      def consume_hackathon(data); end
-
       def consume_place(data)
         consume data['event']
       end
 
+      # TODO: Investigate why these are here, they seem like other types than event?
       def consume_brand(data); end
```
Comment (kimadactyl) [inline app/jobs/calendar_importer/parsers/ld_json.rb:71]: These also seem out of place to me

File: lib/tasks/events.rake [hunk 3]
```diff
@@ -1,23 +1,7 @@
...
-    # scope = scope.where_idle if force_import == false
-
-    scope.find_each do |calendar|
-      puts "queueing #{calendar.name}"
-      calendar.queue_for_import! force_import, from
-    end
-  end
-
-  # rake import:all_events
+namespace :events do
```
Comment (kimadactyl) [inline lib/tasks/events.rake:3]: would it make sense for this to be `calendar` instead?

File: Gemfile [hunk 4]
```diff
@@ -75,14 +75,15 @@ gem 'rack-cors', require: 'rack/cors'
...
 gem 'enumerize'
 gem 'friendly_id'
 # gem 'listen' # needed?
+gem 'invisible_captcha'
 gem 'net-imap', require: false
 gem 'net-pop', require: false
 gem 'net-smtp', require: false
-
-gem 'invisible_captcha'
+gem 'observer'
```
[the remark this answers is not in the document]
Comment (kimadactyl) [reply Gemfile:86]: this is a gem that was already included but im now getting yelled at to include explicitly

File: README.md [hunk 5]
```diff
@@ -67,6 +67,25 @@ Amongst other things, this will create an admin user for you:
...
+```sh
+rails events:import_all_calendars
+```
+
+To import one calendar, run:
+
+```sh
+rails events:import_calendar[100]
+
+# nb: zsh users will need to escape brackets
```
[the remark this answers is not in the document]
Comment (kimadactyl) [reply README.md:83]: this is everyone on a mac as zsh is default now, and it held me up for an hour :)

File: Gemfile [hunk 6]
```diff
@@ -75,14 +75,15 @@ gem 'rack-cors', require: 'rack/cors'
 # Utilities
 gem 'auto_strip_attributes'
 gem 'bootsnap', require: false
+gem 'csv'
```
[the remark this answers is not in the document]
Comment (kimadactyl) [reply Gemfile:78]: they dont, its just cleaning up some console warnings when i run stuff

File: lib/tasks/events.rake [hunk 7]
```diff
@@ -45,49 +29,64 @@ namespace :import do
...
+    from = args[:from] ? to_date(args[:from]) : Date.current.beginning_of_day
     calendar_id = args[:calendar_id]
-    force_import = to_boolean(args[:force_import])
+    force_import = args[:force_import] ? to_boolean(args[:force_import]) : true
 
     CalendarImporterJob.perform_now calendar_id, from, force_import
+  rescue StandardError => e
+    backtrace = e.backtrace[...6]
+    puts "\n"
+    puts "#{e.class}: bad thing: #{e}"
```
[the remark this answers is not in the document]
Comment (kimadactyl) [reply lib/tasks/events.rake:63]: i dont really know what it does i just moved it from elsewhere :)

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (katjam) [reply README.md:83]: Keep it then!

**Comment by kimadactyl** [comment]
Screenshot 2024-09-02 at 19 04 39

verified on staging

**Pull request merged**